File: stacks/score-of-parentheses/score-of-parentheses.java
class Solution {

  public int scoreOfParentheses(String S) {
    Stack<Integer> stack = new Stack<>();
    int score = 0;
    for (char c : S.toCharArray()) {
      if (c == '(') {
        stack.push(score);
        score = 0;
      } else {
        score = stack.pop() + Math.max(score * 2, 1);
      }
    }
    return score;
  }

  public int scoreOfParentheses2(String S) {
    int multiplier = 1;
    int res = 0;

    for (int i = 0; i < S.length() - 1; i++) {
      if (S.charAt(i) == '(') {
        if (S.charAt(i + 1) == ')') {
          res += multiplier;
          i++;
        } else {
          multiplier *= 2;
        }
      } else {
        multiplier /= 2;
      }
    }
    return res;
  }
}

File: stacks/score-of-parentheses/score-of-parentheses.ts
function scoreOfParentheses(S: string): number {
  const stack: number[] = [];

  let score: number = 0;

  for (let i = 0; i < S.length; i++) {
    if (S[i] === '(') {
      stack.push(score);
      score = 0;
    } else {
      score = stack.pop() + Math.max(score * 2, 1);
    }
  }
  return score;
}

function scoreOfParentheses2(S: string): number {
  let mult: number = 1;
  let res: number = 0;

  for (let i = 0; i < S.length - 1; i++) {
    if (S[i] === '(') {
      if (S[i + 1] === ')') {
        res += mult;
        i++;
      } else {
        mult *= 2;
      }
    } else {
      mult /= 2;
    }
  }

  return res;
}
